finish microcosm marketing site

leobauza opened the pull request:
<!--
Please make sure to read the contributing Guidelines:
https://github.com/vigetlabs/microcosm/blob/master/CONTRIBUTING.md
-->

Easier to check out copy changes here: https://deploy-preview-506--inspiring-golick-b2d9e5.netlify.com/

Mostly need any feedback on copy, there is also some new CSS and some React code for the hamburger menu.

**What kind of change does this PR introduce?**

- Documentation

**Does this PR introduce a breaking change?**

- No

Fixes #493 : changing it up to get a first version up, we can revise it later to animate the checkmark and what not

Fixes #492 : Added copy

Fixes #491 : Added a nav

nhunzaker committed:
Add fetching and caching to microcosm-graphql. Flesh out example

nhunzaker committed:
Work in progress

nhunzaker committed:
Make execute blocking

nhunzaker committed:
WIP

nhunzaker committed:
Rethink queries in terms of questions and answers

nhunzaker committed:
Use a separate cache for preparation and resolution

nhunzaker committed:
There is no need for hashing based on object identities

nhunzaker committed:
Make caching configurable

nhunzaker committed:
Start thinking through pagination

nhunzaker committed:
Update changelog

nhunzaker committed:
Make flow happy

nhunzaker committed:
Add changelog for 12.12.1/2

nhunzaker committed:
Upstream repo.parallel fix

We encountered an issue on an app using 12.12.x where we needed the
payloads from `repo.parallel` when it completed. This commit upstreams
that change.

nhunzaker committed:
Bump package.json to 13.0.0-alpha

nhunzaker committed:
Ignore test helpers when collecting code coverage

nhunzaker committed:
Update configuration, simplify some operations

nhunzaker committed:
Unskip an execute test

nhunzaker committed:
Simplify resolvers

nhunzaker committed:
Add web deploy command

nhunzaker committed:
Add slack hook

nhunzaker committed:
Format jest config

nhunzaker committed:
Install AWS CLI in website task

nhunzaker committed:
Web task does not need to attach common workspace

nhunzaker committed:
Add installation caching

nhunzaker committed:
Do not cache

nhunzaker committed:
Execute deploy script through yarn

nhunzaker committed:
Attach workspace to website task

nhunzaker committed:
Repo.parallel can accept objects (#452)

This commit updates repo.parallel such that it can accept plain
JavaScript objects. In this case, the return payload is an object
who's keys match the shape provided.

This also allows generator actions to yield objects

nhunzaker committed:
Add concept of subject, update domain dispatch

nhunzaker committed:
Add custom observable

nhunzaker committed:
Fix repo.parallel example

nhunzaker committed:
Fix publish script

nhunzaker committed:
Do not switch to site

nhunzaker committed:
Do not switch directories

nhunzaker committed:
Switch to the correct directory

nhunzaker committed:
One less step

nhunzaker committed:
Flesh out more usage of observables. Build fails, but it'll fail for a while

nhunzaker committed:
Rework some modules to cut size and fix promise middleware

nhunzaker committed:
Fix some lint errors (but not all of them)

nhunzaker committed:
Fix thunk middleware test

nhunzaker committed:
Observable hooks are autobound, remove uid and cancel

nhunzaker committed:
Get some more tests to pass, wireup more generator actions

nhunzaker committed:
Get history removal tests to pass. Add tree visualization.

nhunzaker committed:
Fix the tree visualization

nhunzaker committed:
Remove extraneous tests. Update promise and toggle tests

nhunzaker committed:
Pass more tests

nhunzaker committed:
Begin thinking about fetching

Commit:
set up structure and JS (#455)

* trial 1 for locked scroll behavior

* trying out two columns layout

* experimenting with react-intersection-observer

* set up IntersectionObserver and have it pull from json file

* restructured base layout for new design and did basic styling

* implementing intersection observer to change text on scroll

* added functionality to change view by button and started refactoring

* tweaked Intersection Observer to run smoother, added third section

* removed unnecessary comments

* remove comment

* ran yarn format

* fixed weird formatting by yarn

* fixed more weird formatting by yarn

* tweaking based on PR comments

* finished refactoring to change classnames on rerender

* refactored to rerender based on one state property only

* changed var name to be more clear

* got rid of unnecessary ids

* ran yarn format

* removed prettier as dev dependency and ran yarn format

* Ch/bg color change (#456)

* Ch/layout (#457)

* Ch/qa scrolling (#458)

* moved polyfill inside componentDidMount

* trying new way to import polyfill

Commit:
ch/design tweaks (#461)

* trying out new design mobile

* tweaked side nav to work on scroll

* added correct file for arrow svg and tweaked padding

* added arrow to graphic

* refactored to export out side nav

* ran yarn format

* refactored to use state to add classname

* refactored to store a map that holds DOM elements

* refactored json file to be used for side nav

nhunzaker committed:
Start consolidating features to achieve stability

Commit:
refactor images, change heading colors on scroll (#464)

* refactored how we pull in images using gatsby docs

* color for text and buttons change on scroll

* added all dummy text. refactored how colors change

* refactored color loops in scss

* small tweaks

* ran yarn format

* add icons, minify images (#465)

* add intro and conclusion sections (#466)

* add browser graphic (#467)

nhunzaker committed:
Move microcosm addons out into microcosm-react

nhunzaker committed:
Format

nhunzaker committed:
Add public path to new site deployment

nhunzaker committed:
Remove workspaces

nhunzaker committed:
No reason to use defaults for one CI job

nhunzaker committed:
Add path prefix

nhunzaker committed:
Do not duplicate path option

nhunzaker committed:
Move prefix paths to microcosm-www-next

Commit:
lazy load images (#469)

* added lazy-sizes lib

* added lazy loader styles and fixed some svg things that were causing console errors

* trying new plugin

* yarn format

* removed other plugin, tweaked scss

nhunzaker committed:
Almost there. Almost all tests pass

nhunzaker committed:
Fix microcosm-react tests

nhunzaker committed:
Update CircleCI command

nhunzaker committed:
Add clean up step. Fix some microcosm-react tests

nhunzaker committed:
Clean up

nhunzaker committed:
Pass through parent option, define state as a getter

nhunzaker committed:
Re-enable fork benchmark

nhunzaker committed:
Clean up

nhunzaker committed:
Remove unused import in microcosm.js

nhunzaker committed:
Update preact tests. Properly build microcosm-preact/react

nhunzaker committed:
Fix lint

nhunzaker committed:
Remove microcosm-react-router

Commit:
Ch/animate text and graphics (#471)

* started refactoring to allow for adding classes on update

* added functionality to fade graphics on scroll

* refactored header to use same fade animation as content

* made small tweaks and ran yarn format

* renamed file

* refactored animation

* tweaked anim timing

* refactored fade thresholds and scss

* refactored how animations are done, animated subheadings and button

* ran yarn format, refactored to use class instancence props

* set up flip animations (#472)

Commit:
Ch/change breakpoint (#473)

* started refactoring to break to mobile at lower point

* refactored so that images resize at smaller breakpoint before going to mobile view

* refactored font size for browser graphic

ccchwang committed:
scss fixes for IE

ccchwang committed:
added section-specific touches to graphic

ccchwang committed:
refactored to take advantage of parent selector

ccchwang committed:
created animtion for checkmark on panel 8

ccchwang committed:
fixed animation for safari

Commit:
Ch/refactor scroll (#475)

* switched from using IntersectionObserver to debounce, fixed bug with footer

* yarn format

* removed unnecessary files

* ran yarn format

* tweaked based on PR comments

* refactored fn to be more clear

nhunzaker committed:
Update microcosm-http tests

nhunzaker committed:
Fix lint issue

nhunzaker committed:
Ignore preact build directory

nhunzaker committed:
Update subject and observable tests. Achieve full coverage for subject

nhunzaker committed:
Unsubscribe is now cancel. Full test coverage of Observable and Subject

nhunzaker committed:
Format

nhunzaker committed:
Merge Microcosm preact and Microcosm react

nhunzaker committed:
Use microcosm-dom in package.json workspace

nhunzaker committed:
Format

nhunzaker committed:
Fix microcosm-http tests

nhunzaker committed:
Fix final efficiency tests. Update bundler

nhunzaker committed:
Fix flow issues

nhunzaker committed:
Consistently return null from tree operations

nhunzaker committed:
Remove eslint cache. Tweak tests after rebase

nhunzaker committed:
Revert bootstrap change to circleci.yml

nhunzaker committed:
Simplify format and lint commands. Remove bin scripts. Add caching

nhunzaker committed:
Add prettierignore file

nhunzaker committed:
Fix lint

nhunzaker committed:
Add bootstrap command

nhunzaker committed:
Update flow, merge microcosm-dom test suite

nhunzaker committed:
Fix lint

nhunzaker committed:
Fix reference to bad jest config in microcosm-dom

nhunzaker committed:
Remove duplicative assignments in GraphQL webpack config

nhunzaker committed:
Fix bad references to addons in microcosm-dev-tools

nhunzaker committed:
Make a few assertion tests dev only

Production builds strip out assertions.

nhunzaker committed:
Move microcosm-dom docs to associated packages. Format docs.

nhunzaker committed:
Add linting to flow config

nhunzaker committed:
Directly reference eslint in lint command

nhunzaker committed:
Remove microcosm-dom docs from microcosm

nhunzaker committed:
Presenter::intercept can have multiple handlers

In order to achieve consistency with effects and domains, the
presenter intercept method can now accept an array of handlers.

nhunzaker committed:
Fix typo in presenter comment

nhunzaker committed:
Upgrade node version on CI

nhunzaker committed:
Remove deploy script. Use Netlify

This commit removes the deploy script for Microcosm's
website. Instead, let's give Netlify a go.

nhunzaker committed:
Remove duplicate bootstrap command

nhunzaker committed:
Clean up Subject, Observable, and History interface

1. Remove `start` callback. This is in compliance with the ES spec.
2. Move Observable.hash to Subject.hash. It returns a Subject.
3. History is now a Subject, eliminating history.updates

nhunzaker committed:
Domains and effects are standalone entities

This commit makes a change to the dispatch process such that domains
and effects own their own dispatch. It does this by creating Effect
and Domain classes.

nhunzaker committed:
Make observable error dev only

nhunzaker committed:
Remove concent of "unstarted". Fix start callback in microcosm-dom.

nhunzaker committed:
Replace package-lock.json with yarn.lock

nhunzaker committed:
Do not setup enzyme when testing microcosm

nhunzaker committed:
Fix lifecycle test

nhunzaker committed:
Fix circular dependency

nhunzaker committed:
Make assertion test dev-only

nhunzaker committed:
Update and prune dependencies

nhunzaker committed:
Move cache-directory to node_modules

nhunzaker committed:
Fix modelWillUpdate reference in changelog

Fixes #454

nhunzaker committed:
Update benchmark build references

Microcosm ships a minified build by default now. This broke the benchmarks.

nhunzaker committed:
Only collect coverage from microcosm src directory

nhunzaker committed:
Inline jest extensions

nhunzaker committed:
Do not format snapshots

nhunzaker committed:
Domains can be defined as observables

This commit allows a domain to be a simple observable. Observable
domains do not have a ledger and can not respond to actions. Their
primary use case is to synchronize state with some other source.

nhunzaker committed:
Remove form-serialize dependency from Microcosm

Now that the microcosm addons have been moved to microcosm-dom, it is
no longer needed.

nhunzaker committed:
Domains are subjects

This commit changes the Domain class such that it extends from the
Subject prototype. Additionally, it configures a toJSON method that
serializes a domain according to history's HEAD state.

This means the `repo.answers` namespace can be removed. You can now
get updates to a domain simply by accessing the domain itself. This
required some Presenter tests to be updated.

nhunzaker committed:
Add Observable::map

This commit adds the `map` method to Observables. This is useful for
transforming a stream of values over time, necessary to fully replace
the old style of building models with Presenters.

nhunzaker committed:
Wrap action-form tests in describe block

nhunzaker committed:
Add value prop to ActionForm

This commit adds a value prop to the ActionForm component of
microcosm-dom. When set, the `value` prop is always passed as the
serialized value of the form. This is intended to better support
controlled forms.

nhunzaker committed:
Add documentation for ActionForm value prop

nhunzaker committed:
Lower max-age cache on build badge.

nhunzaker committed:
Micromanage inception

nhunzaker committed:
Add before/after examples to Micromanage readme

nhunzaker committed:
[micromanage] Add schemas

This commit adds an initial draft of schemas. If an entity is created
without a schema, it fails. Additionally, if a property does not match
the schema, it fails.

I like the strictness, but I also know that it would be a bummer for
an API to change a string or number field, or decide to return null. I
still need to figure out the best approach for that.

One idea is to allow it, with an option to send warnings about schema
mismatches in a backchannel. You'd get an alert about the issue, but a
user wouldn't see an error.

This is probably okay for numbers, strings, and booleans. What about
dates?

nhunzaker committed:
[micromanage] Downgrade version to 0.0.0. Add jest

nhunzaker committed:
[micromanage] Use JSON schemas. Start Collection

ccchwang committed:
updated graphic for intro section

ccchwang committed:
trying out new scroll behaviour/look

ccchwang committed:
tweaking layout for new scroll

ccchwang committed:
added js to stick section to bottom

nhunzaker committed:
[microcosm] Unify effect and domain behavior through agent class

This commit adds a new class to describe the dispatch behavior: an
agent. Agents receive action dispatches from history and can manage
their own internal state.

nhunzaker committed:
Add some extra test coverage for proto. Patch up existing coverage

nhunzaker committed:
Add license to repo root and reference in readme

nhunzaker committed:
update packages in readme

ccchwang committed:
made toggle container responsive

ccchwang committed:
fixed footer color

ccchwang committed:
recalc elem stick upon view change

ccchwang committed:
made sections fade in and out

ccchwang committed:
tweak nav

ccchwang committed:
got rid of slow box shadow

nhunzaker committed:
Increase coverage for microcosm-http

ccchwang committed:
finished styling nav/footer

ccchwang committed:
ran prettier

ccchwang committed:
small tweaks

ccchwang committed:
fixed scss formatting

ccchwang committed:
format

nathanielw committed:
Update CONTRIBUTING with the correct bootstrap command

nathanielw committed:
Document use of yarn workspace

nhunzaker committed:
Update license documents and use consistent file extension

nhunzaker committed:
[microcosm] Add remove data helper. Setting undefined removes props

This commit updates the way `set` works such that it removes a
property if it is undefined. It also adds an `remove` function, which
is an alias for `set(object, prop, undefined)`.

nhunzaker committed:
[microcosm] Refactor data functions

nhunzaker committed:
[microcosm] Rework key path terminology

nhunzaker committed:
[microcosm] Increase test coverage for remove

nhunzaker committed:
[microcosm] Increase coverage on merge utility

nhunzaker committed:
Fix bad reference to moved license file

nhunzaker committed:
Change the way merge works to avoid returning EMPTY_OBJECT

leobauza committed:
finish microcosm marketing site

leobauza committed:
fix unused import error

nhunzaker commented on packages/microcosm-www-next/src/data/index.json:
```diff
@@ -2,64 +2,63 @@
+    "browserText": "So that's a quick recap of Microcosm's players and how they work together. Presenters use Domains to render UIs and create Actions based on user inputs, Domains subscribe to the different states of Actions and update themselves triggering changes in the UI by the Presenter. The Repo oversees all this and keep things organized.<br><br>Under the hood there are loads of performan ...
     "bookend": true
   }
 }
```
This is great! I can't help but wonder if we could store this content as some sort of yaml file or something easier to edit, but we can deal with that later.

nhunzaker commented on packages/microcosm-www-next/src/stylesheets/components/_browser-graphic.scss:
```diff
@@ -83,6 +84,14 @@ $left-padding-lg: rem(40);
     background: linear-gradient(#5385e1 70%, #3a69c0 70%);
+
+    .section.-bg-6 &::after {
+      content: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAsAAAAVCAYAAACQcBTNAAAA0ElEQVQokZXRMUoDURAG4E+0EiVgCAQEK9ucIHZpPEIaLbxACq9g7UFEsAzkAIJWYiCk8gCBRCFYCJJnsW9VllVmp3nM8DE85odzDWqJQRS/4RXHUZwwRyuKE8bYjuKE6yY44aIJ/sBJFCcscBTFCc/YjeKEO2xFccLVf3iFF7znfoPhTuX/U/SwVsS/wR4O8/u9+VER90PuT/863RMOcj/K+LYO36Pzq+/iUxFMp4rbNQsmeftl3fZqnWU8i+B9P2frl8Pq6cpa40aR3KocfgH1Pl0E6oN4JAAAAABJRU5ErkJggg==);
```
We should track an issue to use webpack for this. It would be great if this linked to an asset within the repo and automatically inlined it if it was below a certain size.

nhunzaker reviewed (approved):
It's really exciting to see this come together!

nhunzaker committed:
[microcosm] Remove React dev dependencies

Now that the addons are a separate package, Microcosm no longer
depends on React.

nhunzaker committed:
[microcosm] Patch and Reset only trigger on complete

This commit fixes a bug where patch and reset would trigger on every
state update. I only noticed this after doing some work on a scheduler
for Microcosm, which has some improvements in error handling.

nhunzaker committed:
[microcosm] Do not process actions that are only started

This commit commit adds an optimization to domain dispatch that
prevents the ledger from assigning state for actions that have only
started, but have no progress.

nhunzaker committed:
[microcosm] Update flow types to better handle default state

leobauza committed:
add description, image, and favico

nhunzaker committed:
Fix examples. Fix ReactRouter edge case with Presenters

This commit updates the examples to work with the alpha version of
Microcosm. It also addresses an issue where ReactRouter would not
update due to the use of PureComponent in microcosm-dom.

nhunzaker committed:
Replace console.assert with assert

This commit replaces console.assert with the assert Node standard
library. Jest has decided that console.assert should not raise an
error, blocking us from upgrading to the latest version.

nhunzaker committed:
Add file-uploader example. Updates to microcosm-dom

nhunzaker committed:
Remove unused webpack variable in microcosm-devtools

leobauza committed:
tweak sizing of slides

nhunzaker committed:
Fix typos and rephrase some tests

nhunzaker committed:
Use older version of workspaces pattern

nhunzaker committed:
Update data helpers to fix flow issues

nhunzaker committed:
A few minor fixes to history

- Send pass/fail callbacks to Promise chain in .then()
- Call toString on commands instead of String(commands). It is much
  faster.

nhunzaker committed:
[microcosm] Async observables

This commit updates the Observable abstraction in Microcosm such that
it is asynchronous.

Updates to observables are queued with a universal scheduler. Until
this scheduler "flushes" an observable, updates are buffered within an
internal list of updates.

The scheduler dispatches updates as fast as the browser will allow it
to, using requestIdleCallback. When requestIdleCallback is not
provided, it uses setTimeout.

nhunzaker committed:
[microcosm] Async, scheduled, Subjects

This commit brings the asychronous behavior of Observables over to
Subjects. It is also able to consolidate some behavior by extending
Observable.

The async behavior of Subjects is different in that they track a
present value. When subscribed to, Subjects immediately emit an
update. All future work is scheduled.

This is a really nice system because it allows Presenter view models
to immediately have a value, and for state to be synchronized by the
time Effects dispatch.

nhunzaker committed:
[microcosm] Split Subject.hash into SubjectMap class

The Subject.hash method was getting pretty massive, and I didn't like
having it as a static method on Subject. This commit moves this
function into a separate class that extends from Subject.

nhunzaker committed:
Move cleanup code in observable so that it is closer together

nhunzaker committed:
Move position of incrementer in observer queue

nhunzaker committed:
Fix typo in observable test

nhunzaker committed:
Subscription states should be incremental

nhunzaker committed:
Update tests with asychronous behavior

nhunzaker committed:
Clean up some final items from async work

nhunzaker committed:
Use subscriber interface in Subject

nhunzaker committed:
Use wildcards to avoid circular dependencies

nhunzaker committed:
Drop new keyword in front of Subject.hash

nhunzaker committed:
Fix some edge cases in remove tests for arrays

nhunzaker committed:
Remove rollup-plugin-json

leobauza committed:
add hamburger menu and remove oversized panels

nhunzaker committed:
Update README.md

nhunzaker committed:
Start thinking about the breakdown

nhunzaker committed:
Start fleshing out the cache

nhunzaker committed:
Use proper name for subject-map tests

nhunzaker committed:
[microcosm] Always, synchronously, emit the first observer.next()

This commit updates Observable such that the first "next" callback
invokes immediately. This is important for Observable::map,
particularly when binding data in React. Otherwise, React renders
before the scheduler clears the queue for an observable.

nhunzaker committed:
Add note about scheduler error

nhunzaker committed:
Start thinking through a presentation layer cache

nhunzaker committed:
Fix lint issues

nhunzaker committed:
Update babel-env settings

nhunzaker committed:
Resolve pagination problem

nhunzaker committed:
Resolve pagination problem

nhunzaker committed:
Handle override case

nhunzaker committed:
Add todo note

nhunzaker committed:
Cleanup

nhunzaker committed:
Cleanup micromanage dependencies

nhunzaker committed:
Remove micromanage lockfile

dlederle committed:
Update broken links

The docs directory appears to have moved to the `gh-pages` branch, so I've updated the links. I was unable to find a direct link for `fork` (The link titled 'Split up application state in large apps') but guessed from reading that it was referencing Domains.

nhunzaker committed:
Address dependency issue with microcosm-devtools

Looks like one of the microcosm-devtools dependencies relied on a
behavior in Chrome for sorting that was actually a bug. This commit
updates the usage of that dependency to work around it.

I've also directly targeted Microcosm 12.x until we update 13.x to
support the devtools.

nhunzaker committed:
Bump microcosm devtools to 0.0.4

nhunzaker committed:
Add notes to help with debugging of chrome extension. Tweek text size on tree

nhunzaker committed:
Format devtools example

nhunzaker committed:
Rename connect to query, use Apollo-like interface

nhunzaker committed:
Fix flow type

nhunzaker committed:
Bump microcosm 13.0.0-alpha.0

nhunzaker committed:
Bump microcosm and microcosm-dom to 13.0.0-alpha.1

nhunzaker committed:
Fix build issue in microcosm-dom

nhunzaker committed:
Avoid using Object.assign

nhunzaker committed:
Drop assertion usage

nhunzaker committed:
Bump microcosm-dom to 13.0.0-alpha.2

nhunzaker committed:
Update microcosm-dom readme

nhunzaker committed:
Drop hard dependency on Microcosm in MicrocosmDOM

nhunzaker committed:
Pass next when completing with a payload. Improve efficiency

This commit allows Subjects to be completed with a payload. This
improves ergonomics and compatability with usage in Microcosm 12.x.

The drawback to this approach is that an extra action status is
dispatched. I have alleviated that by adding a `filter` method to the
Observable prototype that Domains use to avoid dispatch if an action
update has no impact on their state.

This actually ended up being a greater efficiency update than I
anticipated, allowing me to unskip a few tests.

nhunzaker committed:
Add back patch/reset API

This commit adds back patch and reset as deprecated API calls. This is
for backwards compatability with Microcosm 12.x.

nhunzaker committed:
Format. Fix flow issues

nhunzaker committed:
Add exceptions for push/reset in flow

nhunzaker committed:
Revert assertion change to function guard

nhunzaker committed:
Use lifecycle constants in Subject

nhunzaker committed:
Add confirmation that filter does not eagerly process iterations

nhunzaker committed:
Add back max history option

This commit adds back the max history option to Microcosm, a feature
of 12.x that controls how many actions Microcosm's history should hold
on to.

This is zero by default in order to prevent memory leaks. However some
apps (like email editors we have built in the past) may desire to hold
on to a set number of actions for undo/redo and reporting features.

nhunzaker committed:
Make domain creation when using observables clearer

nhunzaker committed:
Upgrade build dependencies (#522)

* Upgrade build dependencies
* Bump CircleCI node image to 8.10
* Address minification bugs upgrading Uglify

nhunzaker committed:
[microcosm] Add flow types for scheduler

The scheduler is a critical piece of Microcosm 13, and it didn't have
any flow type coverage.

Adding flow types uncovered a bug where Observable was accessing the
scheduler's error handler from the wrong namespace.

nhunzaker committed:
Testing the scheduler is dev-only

The observable test to confirm proper error reporting uses an internal
API. As such, it must be marked as a dev-only test.

nhunzaker committed:
Pass correct arguments into dev description block

nhunzaker committed:
Handle promise errors using second argument of then() (#524)

This commit changes the way Subjects and Microcosm action resolution
handle promises such that the error state is passed into the second
argument of `promise.then()`.

This prevents a case where callbacks following the success case could
raise an exception, but it would be trapped by the `.catch()` API call.

leobauza committed:
update copy
